arangod/Aql/RegisterPlan.cpp

```diff
@@ -402,22 +393,107 @@ auto RegisterPlanT<T>::clone() -> std::shared_ptr<RegisterPlanT> {
   other->depth = depth;
   other->varInfo = varInfo;
 
-  // No need to clone subQueryNodes because this was only used during
+  // No need to clone subqueryNodes because this was only used during
   // the buildup.
 
   return other;
 }
 
 template <typename T>
 void RegisterPlanT<T>::increaseDepth() {
-  depth++;
+  ++depth;
+
   // create a copy of the last value here
   // this is required because back returns a reference and emplace/push_back
   // may invalidate all references
   auto regCount = nrRegs.back();
   nrRegs.emplace_back(regCount);
 }
 
+template <typename T>
+void RegisterPlanT<T>::shrink(T* start) {
+  // remove all registers greater than maxRegister from the ExecutionNodes 
+  // starting at current that have the same depth as current
+  auto removeRegistersGreaterThan = [this](T* current, RegisterId maxRegister) {
+    auto depth = current->getDepth();
+    while (current != nullptr && current->getDepth() == depth) {
+      current->removeRegistersGreaterThan(maxRegister);
+      current = current->getFirstDependency();
+    }
+  };
+
+  // scan the execution plan upwards, starting at the root node, until
+  // we either find the end of the plan or a change in the node's depth.
+  // while scanning the ExecutionNodes of a certain depth, we keep track 
+  // of the maximum register id they use.
+  // when we are at the end of the plan or at a depth change, we know which
+  // is the maximum register id used by the depth. if it is different from
+  // the already calculated number of registers for this depth, we walk over
+  // all nodes in that depth again and remove the superfluous registers from
+  // them.
+  
+  auto maxUsedRegister = [this](auto const& container, RegisterId maxRegisterId) -> RegisterId {
+    for (auto const& v : container) {
+      auto it = varInfo.find(v->id);
+      if (it != varInfo.end()) {
+        RegisterId rv = it->second.registerId;
+        maxRegisterId = std::max(rv, maxRegisterId);
+      }
+    }
+    return maxRegisterId;
+  };
+  
+  // max RegisterId used by nodes on the current depth
+  RegisterId maxRegisterId = 0;
+  
+  // node at which the current depth starts
+  T* depthStart = start;
+
+  // loop node pointer, will be modified while iterating
+  T* current = start;
+  
+  // currently used variables, outside of loop because the set is recycled
+  VarSet varsUsedHere;
+
+  while (current != nullptr) {
+    // take the node's used registers into account
+    // the maximum used RegisterId for the current depth is tracked in
+    // maxRegisterId
+    varsUsedHere.clear();
+    current->getVariablesUsedHere(varsUsedHere);
+    maxRegisterId = maxUsedRegister(varsUsedHere, maxRegisterId);
+    maxRegisterId = maxUsedRegister(current->getVariablesSetHere(), maxRegisterId);
+    for (auto const& usedLater : current->getVarsUsedLaterStack()) {
+      maxRegisterId = maxUsedRegister(usedLater, maxRegisterId);
+    }
```

Comment:
I'm not quite sure how to make this correct here. It may be easier to do it during `RegisterPlanWalker::after`. There is `unusedRegisters` being maintained, a `std::vector<std::set<RegisterId>>`. That could be used to find rightmost unused registers on every level. Conversely, `regVarMappingStack` should contain, for each level, the variable in each register if any - the maximum register found in there should be exactly the right one for each node.

Reply:
I tried it 2 times to integrate the register tracking into `RegisterPlanWalker::after`, but both attempts failed.
The problem is that you only know that some registers will be unused when at the end of a depth. However, there it is already too late because all nodes for this depth have already been processed. So one would at least go back to all nodes from the same depth and change their settings.
I found that a bit clumsy as well, so I decided to add the shrinking as a completely orthogonal extra step.

Reply:
I don't have any objections to making this an additional step, especially if it makes things easier. What I mean is, this is currently overly pessimistic, for once due to checking against `varsUsedLater` plus `varsSetHere`, rather than a more precise list of currently needed registers. And I don't think we have enough information at this point to do it correctly.

Another thing I've noticed (I've played around a little 😉): It seems that the depth isn't granular enough for this to work well in some cases, and should be increased for more blocks. But maybe we don't want to touch this now, I just wanted to note it.

> However, there it is already too late because all nodes for this depth have already been processed

I think that can be dealt with:

The ExecutionNodes do not get `nrRegs` set, but only `depth`, plus a shared_ptr to the `RegisterPlan` instance. This in turn holds `std::vector<RegisterCount> nrRegs`, which is a map `depth -> nrRegs`. Thus changing the entry in that vector retroactively changes `nrRegs` for all already processed nodes of the same depth.

And neither `regsToClear` nor `regsToKeepStack` may hold a register larger than `nrRegs`, so they must not be changed in any case.

The only thing left I can see is `unusedRegsByNode`, which we can iterate over once in `planRegisters` after the walker is done.

Or am I overlooking something else?